src/zitadel.rs

```diff
@@ -41,434 +58,263 @@ impl Zitadel {
 	}
 
-	/// Update a Zitadel user
-	async fn update_user(&self, old: &ZitadelUser, new: &ZitadelUser) -> Result<()> {
+	/// Import a user into Zitadel
+	pub async fn import_user(&mut self, imported_user: &User) -> Result<()> {
+		tracing::info!("Importing user with external ID: {}", imported_user.external_user_id);
+
 		if self.feature_flags.is_enabled(FeatureFlag::DryRun) {
-			tracing::info!("Not updating user due to dry run: {:?} -> {:?}", old, new);
+			tracing::warn!("Skipping import due to dry run");
 			return Ok(());
 		}
 
-		let Some(user_id) = self.get_user_id(old).await? else {
-			bail!("could not find user `{}` to update", old.user_data.email);
-		};
+		let mut metadata = vec![SetMetadataEntry::new(
+			"localpart".to_owned(),
+			Uuid::new_v5(&FAMEDLY_NAMESPACE, imported_user.external_user_id.as_bytes()).to_string(),
+		)];
 
-		if old.user_data.email != new.user_data.email {
-			self.zitadel_client
-				.update_human_user_name(
-					&self.zitadel_config.organization_id,
-					user_id.clone(),
-					new.user_data.email.clone().to_string(),
-				)
-				.await?;
+		if let Some(preferred_username) = imported_user.preferred_username.clone() {
+			metadata.push(SetMetadataEntry::new(
+				"preferred_username".to_owned(),
+				preferred_username.to_string(),
+			));
+		}
 
-			tracing::warn!(
-				"User email/login changed for {} -> {}",
-				old.user_data.email,
-				new.user_data.email
+		let mut user = AddHumanUserRequest::new(
+			SetHumanProfile::new(
+				imported_user.first_name.to_string(),
+				imported_user.last_name.to_string(),
+			)
+			.with_nick_name(imported_user.external_user_id.to_string())
+			.with_display_name(imported_user.get_display_name()),
+			SetHumanEmail::new(imported_user.email.to_string())
+				.with_is_verified(!self.feature_flags.is_enabled(FeatureFlag::VerifyEmail)),
+		)
+		.with_organization(
+			Organization::new().with_org_id(self.zitadel_config.organization_id.clone()),
+		)
+		.with_metadata(metadata);
+
+		if let Some(phone) = imported_user.phone.clone() {
+			user.set_phone(
+				SetHumanPhone::new()
+					.with_phone(phone.to_string())
+					.with_is_verified(!self.feature_flags.is_enabled(FeatureFlag::VerifyPhone)),
 			);
 		};
 
-		if old.user_data.first_name != new.user_data.first_name
-			|| old.user_data.last_name != new.user_data.last_name
-		{
-			self.zitadel_client
-				.update_human_user_profile(
-					&self.zitadel_config.organization_id,
-					user_id.clone(),
-					new.user_data.first_name.clone().to_string(),
-					new.user_data.last_name.clone().to_string(),
-					None,
-					Some(new.get_display_name()),
-					None,
-					None,
-				)
-				.await?;
-		};
+		if self.feature_flags.is_enabled(FeatureFlag::SsoLogin) {
+			user.set_idp_links(vec![IdpLink::new()
+				.with_user_id(imported_user.external_user_id.to_string())
+				.with_idp_id(self.zitadel_config.idp_id.clone())
+				// TODO: Figure out if this is the correct value; empty is not permitted
```

Comment:
This should probably be confirmed to work before we merge; we have no way to test if IDP logins *actually* work.

Reply:
Since I ended up reading the source code to figure out how Zitadel handles IDP IDs, I also figured out what the "name" means. I originally just wrote this comment:

> This refers to the "username" that the search filter defined in the IDP config resolves to. This means that for our IDP links to work, that needs to be configured to resolve to the user's email address. See: https://github.com/zitadel/zitadel/blob/250f2344c8c2292ca9b861cdd12223d0b4719d43/internal/idp/providers/ldap/session.go#L230

That should make it into the readme, though.